drivers/storage/portworx/deployment.go

```diff
@@ -318,11 +320,46 @@ func (p *portworx) GetStoragePodSpec(
 	}
 
 	if cluster.Spec.CloudStorage != nil && len(cluster.Spec.CloudStorage.CapacitySpecs) > 0 {
-		// Cloud Environment
-		cloudConfig, err := p.generateCloudStorageSpecs(cluster)
+		var cloudConfig *cloudstorage.Config
+		storageNodes, err := p.getStorageNodes()
 		if err != nil {
 			return v1.PodSpec{}, err
 		}
+		if len(storageNodes.Items) == 0 {
+			cloudConfig, err = p.generateCloudStorageSpecs(cluster)
+			if err != nil {
+				logrus.Error(err)
+				return v1.PodSpec{}, err
+			}
+
+			nodes, err := p.getNodesList()
+			if err != nil {
+				return v1.PodSpec{}, err
+			}
+
+			for _, node := range nodes.Items {
```

Comment:
We should generate the StorageNode object only for one node here and not for all the nodes. The list returned by getNodesList contains all the nodes in the kubernetes cluster, but that does not mean storage provider (or Portworx) would be running on all of them. 

May be pass in the node name in GetStoragePodSpec() and only create StorageNode object for that node.

Reply:
I thought the idea was to keep a separate spec for each storage node, wasn't it?

Reply:
Yes we want to create a StorageNode object only for those nodes where Portworx will be running. Lets say a user has provided node affinity rules where in Portworx needs to be installed on node1, node2, node3 and skip node4, node5.

The controller makes sure that Portworx pod does not start on node4 and node5. But this for loop will create 5 StorageNode objects. 

Which is why I am suggesting pass in the NodeName in GetStoragePodSpec and only create the StorageNode object for that node.